compiler/rustc_const_eval/src/interpret/step.rs

```diff
@@ -157,26 +147,29 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         rvalue: &mir::Rvalue<'tcx>,
         place: mir::Place<'tcx>,
     ) -> InterpResult<'tcx> {
-        let dest = self.eval_place(place)?;
-
         use rustc_middle::mir::Rvalue::*;
         match *rvalue {
             ThreadLocalRef(did) => {
+                let dest = self.eval_place(place)?;
                 let ptr = M::thread_local_static_base_pointer(self, did)?;
                 self.write_pointer(ptr, &dest)?;
             }
 
             Use(ref operand) => {
+                // FIXME: Here, in `Aggregate`, and in `Repeat`, the various `dest = eval_place`s
+                // should also be retagging the pointer to ensure that it does not overlap with the
+                // RHS.
+                let dest = self.eval_place(place)?;
                 // Avoid recomputing the layout
                 let op = self.eval_operand(operand, Some(dest.layout))?;
                 self.copy_op(&op, &dest)?;
             }
 
             BinaryOp(bin_op, box (ref left, ref right)) => {
-                let layout = binop_left_homogeneous(bin_op).then_some(dest.layout);
-                let left = self.read_immediate(&self.eval_operand(left, layout)?)?;
+                let left = self.read_immediate(&self.eval_operand(left, None)?)?;
```

Comment:
This needs careful benchmarking; the perf gain from that optimization was quite noticeable in the past.